restapi/src/main/java/io/stargate/web/docsapi/service/query/DocumentSearchService.java

```diff
@@ -0,0 +1,374 @@
+    }
+
+    // get candidates that pass the cassandra conditions
+    // note that this returns empty flowable if no cassandra conditions are found
+    return getCandidates(db, keyspace, collection, expression, paginator, null)
+
+        // TODO DO NOT BREAK THE CHAIN, losing the page state in the original doc
+        // map to doc id to dismiss unusable rows
+        .map(RawDocument::id)
+
+        // filter in memory
+        .concatMap(filterInMemory(db, keyspace, collection, expression, paginator))
+
+        // and then populate the docs that passed with the requested fields
+        .concatMap(populateDocument(db, keyspace, collection, fields))
+
+        // if empty either no document is found, or we had no cassandra conditions
+        .switchIfEmpty(
+            Single.fromCallable(() -> hasPersistenceConditions(expression))
+
+                //
+                .flatMapPublisher(
+                    hadPersistence -> {
+
+                      // if we had persistence, then no docs found, break
+                      if (hadPersistence) {
+                        return Flowable.empty();
+                      }
+
+                      // TODO or start from the in memory filters directly if we have them
+                      //  Dmitri >>> not sure how would this work from the paging point of view?
+
+                      // otherwise loaf full document and test for the in memory stuff
+                      // do a non-predicated search based on the paging state
+                      return getAllByPagingState(db, keyspace, collection, paginator)
+
+                          // then test for the in-memory conditions
+                          // no need to populate, fetched whole docs already
+                          .concatMap(testInMemoryConditions(expression));
+                    }));
+  }
+
+  private Flowable<RawDocument> getCandidates(
+      DocumentDB db,
+      String keyspace,
+      String collection,
+      Expression<FilterExpression> expression,
+      Paginator paginator,
+      Flowable<RawDocument> documents) {
+
+    // start from expression
+    return Single.just(expression)
+        .flatMapPublisher(
+            e -> {
+
+              // resolve if we have remaining persistence stuff
+              boolean hasPersistence = hasPersistenceConditions(e);
+
+              // if so continue and find next persistence expression to resolve
+              if (hasPersistence) {
+                return nextPersistenceExpressions(e)
+                    .flatMapPublisher(
+                        groupExpressions -> {
+
+                          // simplify expression considering that expressions would be solved on C*
+                          // TODO this is bad, we need to simplify the expression not only remove
+                          // from nors
+                          Expression<FilterExpression> afterQueryExpression =
+                              removeExpressions(e, groupExpressions);
```

Comment:
So something like this is what we need to figure out.. This works well in case of:

`A & B`

But as soon as you have the OR, it will not work.. Consider:

`C1 & (C2 | M1)`

where `C` is a C* filter and `M` is a memory filter.. You can start with A, but then you can not follow with B I guess.. However, if we can look to `C2 | M1` as a single expression (having a special logic to resolve it), then it would work, right? 

Let's shortly talk about this..

Reply:
As agreed.